File: script/generator.js
/**
 * ---------------------------------------------------------------------------
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ---------------------------------------------------------------------------
 */

const fs = require('fs');
const xlsx = require('xlsx');
const path = require('path');
const tracks_dictionary = require('./tracks_dictionary.json');
const { console } = require('inspector');

const outputDir = path.join(__dirname, '../content', 'sessions');
if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir, { recursive: true });
}

// 读取Excel文件中的会议session数据
function readExcelFile(filePath) {
  const workbook = xlsx.readFile(filePath);
  const sheetName = workbook.SheetNames[0];
  const worksheet = workbook.Sheets[sheetName];
  // 将工作表转换为JSON对象数组
  return xlsx.utils.sheet_to_json(worksheet);
}

// 定义session类型的英文映射字典
const sessionTypesDictionary = {
  "Chinese": "Chinese Session",
  "English": "English Session"
};

// 定义session类型的中文映射字典
const sessionTypesChineseDictionary = {
  "Chinese": "中文演讲",
  "English": "英文演讲"
};

function formatExcelDate(excelDate) {
  if (!excelDate) return '';
  const date = new Date(Math.round((excelDate - 25569) * 86400 * 1000));
  return date.toISOString().slice(0, 19);
}

function updateFileContent(filePath, newFrontmatter, newContent) {
  if (fs.existsSync(filePath)) {
    console.log('File already exists skip');
  } else {
    const fullContent = `---\n${newFrontmatter}\n---\n\n${newContent}`;
    fs.writeFileSync(filePath, fullContent, 'utf8');
  }
}


function processAndGroupSessions(sessions) {
  const sessionMap = new Map();
  
  sessions.forEach(session => {
    const sessionId = session['Session Id'];
    if (!sessionId) return;
    
    if (!sessionMap.has(sessionId)) {
      // 首次遇到该 Session Id，创建基础数据
      sessionMap.set(sessionId, {
        ...session,
        speakers: [], // 存储所有讲师信息对象数组
      });
    }
    
    // 获取当前存储的数据
    const existingData = sessionMap.get(sessionId);
    
    // 构建讲师信息对象
    if (session['FirstName'] || session['LastName']) {
      const speakerInfo = {
        name: `${session['FirstName'] || ''} ${session['LastName'] || ''}`.trim(),
        avatar: `<img src="${session['Profile Picture']}" width="200" /><br/>\n` || '',
        bio: session['Bio'] || '',
        tagline: session['TagLine'] || '',
      };
      
      // 检查是否已存在相同讲师（通过名字判断）
      const existingSpeaker = existingData.speakers.find(s => s.name === speakerInfo.name);
      if (!existingSpeaker) {
        existingData.speakers.push(speakerInfo);
      }
    }
  });
  
  // 处理合并后的数据
  return Array.from(sessionMap.values()).map(session => ({
    ...session,
    // speakers数组已经是我们需要的格式，不需要额外处理
    speakers: session.speakers
  }));
}


/**
 * 主函数 - 处理Excel文件并生成Markdown文件
 */
function main() {
  const rawSessions = readExcelFile('./ApacheConSessions.xlsx');
  
  // 处理和合并数据
  const processedSessions = processAndGroupSessions(rawSessions);
  
  // 遍历所有session数据
  for (let index = 0; index < processedSessions.length; index++) {
    const session = processedSessions[index];
    
    // 只处理状态为"accept"的被接受的session
    // if (session['Status'] === 'Accepted') {
      const track = tracks_dictionary[session['Track']];
      const fileName = `${track}-${session['Session Id']}`;
      const enFilePath = path.join(outputDir, `${fileName}.md`);
      const zhFilePath = path.join(outputDir, `${fileName}.zh.md`);
      // 获取日期信息
      let scheduleTime = "";
      // if (session['Date Submitted']) {
      //   scheduleTime = formatExcelDate(session['Date Submitted']);
      // }
      
      // 获取英文标题
      let title = session['Title'];
      
      // 映射session类型
      const sessionType = sessionTypesDictionary[session['Language']];
    
      const speakers = session['speakers'].map(v => v.name).join(', ');
      
      // 获取英文摘要和演讲者简介
      const abstract = session['Description'];
      let speakerBios = '';
      session['speakers'].forEach(speaker => {
        speakerBios += `\n${speaker.avatar}\n${speaker.name}: ${speaker.tagline}\n\n${speaker.bio}\n\n`;
      })
      
      // 修改后:
      const enFrontmatter = `title: "${title}"
date: "${scheduleTime}"
track: "${track}"
presenters: "${speakers}"
stype: "${sessionType}"`;
      const enContent = `${abstract}\n\n### Speakers:\n\n${speakerBios}`;
      updateFileContent(enFilePath, enFrontmatter, enContent);


      // .zh.md
      const sessionZhType = sessionTypesChineseDictionary[session['Language']];
      const zhFrontmatter = `title: "${title}"
date: "${scheduleTime}"
track: "${track}"
presenters: "${speakers}"
stype: "${sessionZhType}"`;
        const zhContent = `${abstract}\n\n### 讲师:\n\n${speakerBios}`;
      updateFileContent(zhFilePath, zhFrontmatter, zhContent);
    // }
  }
}


main()

File: script/generator.py
import os
import pandas as pd

print("开始执行脚本...")
# 从Excel文件读取会议session数据
sessions = pd.read_excel('ApacheConSessions.xlsx')
print(f"Excel文件读取完成，共有{len(sessions)}行数据")

# 确保目标目录存在
output_dir = "content/sessions"
if not os.path.exists(output_dir):
    os.makedirs(output_dir)
    print(f"创建目录: {output_dir}")

# 定义track(轨道/分类)的映射字典，将中文/英文分类名映射为URL友好的格式
tracks_dictionary = {"Cloud Native": "cloudnative",
                     "DataOpts": "dataopts",
                     "Streaming": "streaming",
                     "Incubator": "incubator",
                     "Data Lake & Data Warehouse": "datalake",
                     "OLAP & Data Analysis": "olap",
                     "Community": "community",
                     "Data Storage & Computing": "datastorage",
                     "Web Application & Framework": "webapplication",
                     "AI": "ai",
                     "IoT": "iot",
                     "Message": "message",
                     "Microservice": "microservice",
                     "Observability": "observability",
                     "Rust": "rust",
                     "General": "general",
                     "Keynote": "keynote",
                     "5-mins-Lightning-Talk": "5minslightningtalk",
                     }
# 定义session类型的英文映射字典
session_types_dictionary = {"Chinese": "Chinese Session",
                            "English": "English Session"}
# 定义session类型的中文映射字典
session_types_chinese_dictionary = {"中文演讲": "中文演讲",
                                    "Regular Session": "英文演讲"}

# 移除演讲者信息中的email地址(如果有)
def remove_speakers_email(speakers):
    if pd.isnull(speakers):
        return ""
    speaker_list = speakers.split(",") # 将多个演讲者按逗号分隔
    result_list = []
    for speaker in speaker_list:
        result_list.append(remove_speaker_email(speaker)) # 对每个演讲者移除email
    return ', '.join(result_list) # 重新组合为字符串

# 移除单个演讲者的email地址(格式通常为"姓名 <email>")
# 取<符号前的部分并去除空格
def remove_speaker_email(speaker):
    parts = speaker.split('<')
    if len(parts) > 1:
        return parts[0].strip()
    return speaker.strip()

# 遍历所有session数据
for index in sessions.index:
    # 只处理状态为"Accepted"的被接受的session
    if sessions.at[index, 'Status'] == 'Accepted':
        # create the markdown file
        
        # 获取track值
        track_value = tracks_dictionary.get(sessions.at[index, 'Track'], "unknown")
        session_id = str(sessions.at[index, 'Session Id'])
        
        # 生成文件名: Track-Session Id
        file_name = f"{track_value}-{session_id}"
        
        # 创建英文和中文的markdown文件路径
        en_file_path = os.path.join(output_dir, file_name + ".md")
        zh_file_path = os.path.join(output_dir, file_name + ".zh.md")
        
        
        # 创建英文和中文的markdown文件
        markdown_en_file = open(en_file_path, "w+", encoding="utf-8")
        markdown_zh_file = open(zh_file_path, "w+", encoding="utf-8")
        
        # markdown文件的元数据模板
        mate_data = '---\ntitle: "{}"\ndate: "{}" \ntrack: "{}"\npresenters: "{}"\nstype: "{}"\n---\n'
        schedule_time = ""
        if not pd.isnull(sessions.at[index, 'Date Submitted']):
            schedule_time = sessions.at[index, 'Date Submitted']
        # 获取英文标题
        title = sessions.at[index, 'Title']

        # map the session type
        session_type = session_types_dictionary.get(sessions.at[index, 'Language'])

        # update the track information
        # 映射track分类到URL友好格式
        track = track_value
        # 处理演讲者信息(移除email)
        speakers = remove_speakers_email(sessions.at[index, 'Owner'])
        # 获取英文摘要和演讲者简介
        abstract = sessions.at[index, "Description"] if not pd.isnull(sessions.at[index, "Description"]) else ""
        speaker_bios = sessions.at[index, "Bio"] if not pd.isnull(sessions.at[index, "Bio"]) else ""
        
        speaker_image = ""
        if not pd.isnull(sessions.at[index, 'Profile Picture']):
            speaker_image = sessions.at[index, 'Profile Picture']
            head_img_html = f'<img src="images/speaker/{speaker_image}" width="200" />'
        else:
            head_img_html = '' 
        
        # 写入英文markdown文件
        markdown_en_file.write(mate_data.format(title,
                                               schedule_time,
                                               track,
                                               speakers,
                                               session_type))
        markdown_en_file.write(abstract)
        markdown_en_file.write("\n ### Speakers: \n ")
        if head_img_html:
            markdown_en_file.write(head_img_html + "\n\n")
        markdown_en_file.write(speaker_bios)
        markdown_en_file.close()
        
        # 处理中文内容(如果存在)
        if not pd.isnull(sessions.at[index, 'Chinese Title']):
            title = sessions.at[index, 'Chinese Title']
        if not pd.isnull(sessions.at[index, "Chinese Abstract"]):
            abstract = sessions.at[index, "Chinese Abstract"]
        if not pd.isnull(sessions.at[index, "Chinese Bios"]):
            speaker_bios = sessions.at[index, "Chinese Bios"]

        # map the session type
        session_zh_type = session_types_chinese_dictionary.get(sessions.at[index, 'Language'])

        markdown_zh_file.write(mate_data.format(title,
                                               schedule_time,
                                               track,
                                               speakers,
                                               session_zh_type))
        markdown_zh_file.write(abstract)
        markdown_zh_file.write("\n ### 讲师: \n ")
        if head_img_html:
            markdown_zh_file.write(head_img_html + "\n\n")
        markdown_zh_file.write(speaker_bios)
        markdown_zh_file.close()
